### PR TITLE
Tensor slice functions and unit tests

### PR DESCRIPTION
**Pull request by jwa7**
**Overview**

Functions to slice equistore tensors based on some specified samples, properties, or both. All functions are documented, and there is a suite of associated unit tests.

This pull request supersedes that of PR #37. These slice functions have been implemented as a helper module in the `operations/` subdirectory as opposed to as a class method in `tensor.py`. 

**Functionality**

In `slice.py`, there are 3 new functions. All are documented and have unit tests.

**1. `_slice_tensorblock(tensorblock, samples_to_slice=None, properties_to_slice=None)`**

* This slices an individual `TensorBlock`, including gradients, according to the arguments `samples_to_slice` and `properties_to_slice`. It returns a new `TensorBlock`, where the shape of the values tensor (`TensorBlock.values`) is smaller than (if slicing has occurred) or equal to (if no slicing has occurred) the input `TensorBlock`.

* `samples_to_slice` and `properties_to_slice` must be passed as `Labels` objects, specifying the name(s) and values of the samples and/or properties to slice by.

* Note that the values tensor could be empty if none of the specified samples and/or properties belong to the input `TensorBlock`. In this case, although there will be no actual values in the `TensorBlock.values` array, the shape of this tensor will be non-zero in the dimensions that haven't been sliced. This is created by slicing the input `TensorBlock`, as opposed to just returning an artificially-created empty one (with no shape or dimensions), and is intentional. It allows the slicing of dimensions to be tracked. 

* For example, if a `TensorBlock` of shape `(52, 1, 5)` is passed, and only some samples are specified to be sliced but none of these appear in the input `TesnorBlock`, the returned `TensorBlock` values array will be empty, but its shape will be `(0, 1, 5)` - i.e. the samples dimension has been sliced to zero but the components and properties dimensions remain in-tact. The same logic applies to any `Gradient TensorBlocks` the input `TensorBlock` may have associated with it.

**2. `_slice_tensormap(tensormap, samples_to_slice=None, properties_to_slice=None)`**

* This works by iterating over and slicing each block of the input `TensorMap` and returning a new `TensorMap`. 

* The returned `TensorMap` will have the same number of blocks, indexed by the same keys, as the input, but some (or all) might be empty, depending on the samples or properties the user has chosen to slice by.

**3. `slice(tensor, samples_to_slice=None, properties_to_slice=None)`**

* This is the user-facing function. It identifies the type of the input tensor (either a `TensorBlock` or a `TensorMap`) and calls the appropriate function from the 2 above. 

* It also handles input checking, and warns users if their returned tensor is empty (in the case of the user passing a `TensorBlock`) or if it contains any (or all) empty blocks (in the case of the user passing a `TensorMap`.

**Outlook**

The split function is useful to allow users to split up their data along the samples or properties dimensions, or both. It works by explicit specification of samples and/or properties. However, on top of this module could be built a 'train_test_split' module where tensors are sliced randomly without the need to explicitly specify the exact indices of the samples/properties they want to keep (which can be tedious). This could also be extended even further to include a function to split data according to a k-fold cross validation scheme. Using the split function, both would be relatively easy to implement and generalise.

### CODE REVIEW

File: python/tests/operations/slice.py [hunk 1]
```diff
@@ -0,0 +1,696 @@
...
+
+import equistore.io
+from equistore.operations import slice, remove_gradients
+from equistore import Labels, TensorBlock, TensorMap
+
+DATA_ROOT = os.path.join(os.path.dirname(__file__), "..", "data")
+TEST_FILE = "qm7-spherical-expansion.npz"
+
+
+class TestSlice(unittest.TestCase):
```
Comment (DavideTisi) [inline python/tests/operations/slice.py:14]: Is there a test in which you are slicing both on samples and properties?

File: python/src/equistore/operations/slice.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                        [[0, 1], [0, 6], [1, 6], [3, 16]]
+                    ),  # must be a 2D-array
+                )
+                properties=Labels(
+                    names=["n",],  # radial channel
+                    values=np.array([[3], [4], [5]]),
+                )
+            )
+    """
+
```
[the remark this answers is not in the document]
Comment (jwa7) [reply python/src/equistore/operations/slice.py:180]: Good point. Change made in commit 49f5043579563816067d9b85d5a82309d604c42f.

File: python/tests/operations/slice.py (hunk 1 above)
Comment (jwa7) [reply python/tests/operations/slice.py:14]: Yes it does, see test block 5. In docstring on TestSlice: "5) Slicing samples and properties dimensions simultaneously."

File: python/src/equistore/operations/slice.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jwa7) [reply python/src/equistore/operations/slice.py:182]: Pushed changed but build failed. Fixed in commit dc69122f417a88a00b7ed5b844929f6a2e0ff3f2.

File: python/src/equistore/operations/slice.py [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        samples=tensorblock.samples[samples_filter],
+        components=tensorblock.components,
+        properties=tensorblock.properties[properties_filter],
+    )
+
+    # Create a list of numeric indices of the samples that were sliced. Needed to
+    # create filters for Gradient TensorBlocks as there isn't a one-to-one mapping
+    # between samples of the parent TensorBlock and its Gradients. This doesn't
+    # need to be done for properties as the mapping is by definition a one-to-one.
+    numeric_idxs = np.array([i for i, val in enumerate(samples_filter) if val])
```
[the remark this answers is not in the document]
Comment (jwa7) [reply python/src/equistore/operations/slice.py:221]: Noted, and remedy attempted, but problems are caused with tensor slicing when trying to slice with an empty array in the case of there being none of the specified samples/properties present in the TensorBlock. For instance, attempted to slice with a length zero numpy array of indices as opposed a length N numpy array of all False values. 

To make this change work one would need to introduce some extra conditionals to `_slice_tensorblock`, which I would argue isn't worth (at least in terms of readability and generalisability) the cost savings from not performing the one extra loop of the samples of the block to create the `numeric_idxs` in addition to the `samples_filter`.

[the remark this answers is not in the document]
Comment (DavideTisi) [reply python/src/equistore/operations/slice.py:221]: for me it is also fine to leave as it is with both the bool and numerical array and simply re do the loop.

File: python/src/equistore/operations/slice.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+            if isinstance(tensor, TensorMap)
+            else tensor.properties.names
+        )
+        for name in properties_to_slice.names:
+            assert name in property_names, (
+                "Invalid property name '" + name + "' - doesn't exist in input tensor."
+            )
+
+    # Slice tensor and issue warnings if some or all blocks are empty
+    warnings.filterwarnings(action="always")
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:97]: Why is this here? We should not be changing the global warnings configuration from inside a library function

File: python/tests/operations/slice.py [hunk 5]
```diff
@@ -0,0 +1,768 @@
...
+        )
+        # Remove the gradients to simplify test
+        tensormap = remove_gradients(tensormap)
+        # Slice only 'n' (i.e. radial channels) -1 (i.e. non-existent channel)
+        channels_to_keep = np.array([-1]).reshape(-1, 1)
+        properties_to_slice = Labels(
+            names=["n"],
+            values=channels_to_keep,
+        )
+        with self.assertWarns(Warning):
```
Comment (Luthaf) [inline python/tests/operations/slice.py:759]: you can also check the warnings message. Check for tests with assertRaises for similar example

**Review by Luthaf — COMMENTED** [review]
Looks very good overall! 

The main change I would suggest is to have two separate functions for slicing blocks/full tensors instead of a single one with `if isinstance ...`. 

Most of the other operations already work this way, and I think it makes it clearer what the code is doing at any point. 

I'll do another quick round of review checking corner cases later today/tomorrow!

**Comment by jwa7** [comment]
> Looks very good overall!
> 
> The main change I would suggest is to have two separate functions for slicing blocks/full tensors instead of a single one with `if isinstance ...`.
> 
> Most of the other operations already work this way, and I think it makes it clearer what the code is doing at any point.
> 
> I'll do another quick round of review checking corner cases later today/tomorrow!

Fixed in commit a6c8fc99604993c52c28564c24b3065430ae7f91. User-facing fxns separated out into `slice()`, which slices just `TensorMap` objects, and `slice_block()`, which slices just `TensorBlock` objects. Most of the functionality remains in the hidden methods `_slice_tensormap()` and `_slice_tensorblock` and thus many of the tests remain the same.

The next commits will address the concerns you raised regarding warnings and how they are tested.

File: python/src/equistore/operations/slice.py (hunk 4 above)
Comment (jwa7) [reply python/src/equistore/operations/slice.py]: Removed in commit f49fe0b215ca61158597a18c8eb92fffacbac168.

File: python/tests/operations/slice.py (hunk 5 above)
Comment (jwa7) [reply python/tests/operations/slice.py]: Done in commit 434abeaa6a8d1b532e8ee011926cb630b336c4da.

**Comment by jwa7** [comment]
All above concerns resolved

File: python/src/equistore/operations/slice.py [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+
+from ..tensor import TensorMap
+from ..block import TensorBlock
+from ..labels import Labels
+
+
+def slice(tensormap: TensorMap, samples_to_slice=None, properties_to_slice=None):
+    """
+    Slices an input :py:class:`TensorMap` along the samples and/or properties
+    dimension(s). `samples_to_slice` and `properties_to_slice` are
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:13]:
```suggestion
    dimension(s). ``samples_to_slice`` and ``properties_to_slice`` are
```

RestructuredText syntax is different from markdown unfortunately

File: python/src/equistore/operations/slice.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+    :py:class:`Labels` objects that specify the samples/properties
+    (respectively) names and indices that should be sliced, i.e. kept in the
+    output tensor.
+
+    Note that either `samples_to_slice` or `properties_to_slice`, or both,
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:18]:
```suggestion
    Note that either ``samples_to_slice`` or ``properties_to_slice``, or both,
```

File: python/src/equistore/operations/slice.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+            properties=Labels(
+                names=["n",],  # radial channel
+                values=np.array([[3,], [4,], [5,]]),
+            )
+        )
+
+    Also note that this function will return a :py:class:`TensorMap` whose
+    blocks are of equal or smaller dimensions (due to slicing) than those of the
+    input. However, the returned :py:class:`TensorMap` will be returned with the
+    same number of blocks and the corresponding keys as in the input. If any
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:40]:
```suggestion
    same number of blocks and the corresponding keys as the input. If any
```

File: python/src/equistore/operations/slice.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+    """
+    Slices an input :py:class:`TensorMap` along the samples and/or properties
+    dimension(s). `samples_to_slice` and `properties_to_slice` are
+    :py:class:`Labels` objects that specify the samples/properties
+    (respectively) names and indices that should be sliced, i.e. kept in the
+    output tensor.
+
+    Note that either `samples_to_slice` or `properties_to_slice`, or both,
+    should be specified as input. User warnings will be issued if any returned
+    blocks are empty after slicing.
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:20]:
```suggestion
    should be specified as input.
```

this is already mentioned below

File: python/src/equistore/operations/slice.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+    blocks are of equal or smaller dimensions (due to slicing) than those of the
+    input. However, the returned :py:class:`TensorMap` will be returned with the
+    same number of blocks and the corresponding keys as in the input. If any
+    block upon slicing is reduced to nothing, i.e. in the case that it has none
+    of the specified `samples_to_slice` or `properties_to_slice`, an empty block
+    will be returned but will still be accessible by its key. User warnings will
+    be issued if any blocks are sliced to contain no values.
+
+    For the empty blocks that may be returned, although there will be no actual
+    values in its TensorBlock.values tensor, the shape of this tensor will be
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:47]:
```suggestion
    values in its ``TensorBlock.values`` array, the shape of this array will be
```

File: python/src/equistore/operations/slice.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+    non-zero in the dimensions that haven't been sliced. This is created by
+    slicing the input TensorBlock, as opposed to just returning an
+    artificially-created empty one (with no shape or dimensions), and is
+    intentional. It allows the slicing of dimensions to be tracked.
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:51]:
```suggestion
    non-zero in the dimensions that haven't been sliced. This allow slicing of dimensions to be tracked.
```

This part of the docs feels a bit verbose, how about cutting them down a bit?

File: python/src/equistore/operations/slice.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+
+    For example, if a TensorBlock of shape (52, 1, 5) is passed, and only some
+    samples are specified to be sliced but none of these appear in the input
+    TesnorBlock, the returned TensorBlock values array will be empty, but its
+    shape will be (0, 1, 5) - i.e. the samples dimension has been sliced to zero
+    but the components and properties dimensions remain in-tact. The same logic
+    applies to any Gradient TensorBlocks the input TensorBlock may have
+    associated with it.
+
+    See the documentation for the `slice_block()` to see how an individual
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:61]:
```suggestion
    See the documentation for the :py:func:`slice_block` to see how an individual
```

File: python/src/equistore/operations/slice.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+    :py:class:`TensorBlock` is sliced.
+
+    :param tensor: the input :py:class:`TensorMap` to be sliced.
+    :param samples_to_slice: a :py:class:`Labels` object containing the names
+        and indices of samples to keep in the sliced :py:class:`TensorBlock`
+        output, or each of the sliced :py:class:`TensorBlock` objects of the
+        output :py:class:`TensorMap`. Default value of None indicates no slicing
+        along the samples dimension should occur.
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:69]:
```suggestion
    :param samples_to_slice: a :py:class:`Labels` object containing the names
        and indices of samples to keep in each of the sliced :py:class:`TensorBlock` 
        of the output :py:class:`TensorMap`. Default value of None indicates no slicing
        along the samples dimension should occur.
```

File: python/src/equistore/operations/slice.py [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+    :param properties_to_slice: a :py:class:`Labels` object containing the names
+        and indices of properties to keep in the sliced :py:class:`TensorBlock`
+        output, or each of the sliced :py:class:`TensorBlock` objects of the
+        output :py:class:`TensorMap`. Default value of None indicates no slicing
+        along the properties dimension should occur.
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:74]:
```suggestion
    :param properties_to_slice: a :py:class:`Labels` object containing the names
        and indices of properties to keep in each of the sliced :py:class:`TensorBlock` 
        of the output :py:class:`TensorMap`. Default value of None indicates no slicing
        along the properties dimension should occur.
```

File: python/src/equistore/operations/slice.py [hunk 15]
```diff
@@ -0,0 +1,395 @@
...
+        + " use the `slice_block()` function instead."
+    )
+    if samples_to_slice is None and properties_to_slice is None:
+        raise ValueError(
+            "Must specify either some samples or properties (or both) to slice by."
+        )
+    if samples_to_slice is not None:
+        assert isinstance(
+            samples_to_slice, Labels
+        ), "samples_to_slice must be passed as a equistore `Labels` object."
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:92]: This would be better as a TypeError (users can disable asserts, and the error message is clearer with a TypeError)

File: python/src/equistore/operations/slice.py [hunk 16]
```diff
@@ -0,0 +1,395 @@
...
+        sample_names = (
+            tensormap.sample_names
+            if isinstance(tensormap, TensorMap)
+            else tensormap.samples.names
+        )
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:97]: this still handles the case of a TensorBlock as input

File: python/src/equistore/operations/slice.py [hunk 17]
```diff
@@ -0,0 +1,395 @@
...
+        for name in samples_to_slice.names:
+            assert name in sample_names, (
+                "Invalid sample name '"
+                + name
+                + "' - doesn't exist in input `TensorMap`."
+            )
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:103]: should be a ValueError

File: python/src/equistore/operations/slice.py [hunk 18]
```diff
@@ -0,0 +1,395 @@
...
+    if properties_to_slice is not None:
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:104]: same as above

File: python/src/equistore/operations/slice.py [hunk 19]
```diff
@@ -0,0 +1,395 @@
...
+
+
+def slice_block(
+    tensorblock: TensorBlock,
+    samples_to_slice=None,
+    properties_to_slice=None,
+) -> TensorBlock:
+    """
+    Slices an input :py:class:`TensorBlock` along the samples and/or properties
+    dimension(s). `samples_to_slice` and `properties_to_slice` are
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:158]:
```suggestion
    dimension(s). ``samples_to_slice`` and ``properties_to_slice`` are
```

File: python/src/equistore/operations/slice.py [hunk 20]
```diff
@@ -0,0 +1,395 @@
...
+    :py:class:`Labels` objects that specify the samples/properties
+    (respectively) names and indices that should be sliced, i.e. kept in the
+    output :py:class:`TensorBlock`.
+
+    Note that either `samples_to_slice` or `properties_to_slice`, or both,
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:163]:
```suggestion
    Note that either ``samples_to_slice`` or ``properties_to_slice``, or both,
```

File: python/src/equistore/operations/slice.py [hunk 21]
```diff
@@ -0,0 +1,395 @@
...
+    should be specified as input. User warnings will be issued if the block is
+    empty after slicing.
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:165]:
```suggestion
    should be specified as input.
```

And similar changes to the other functions

**Review by Luthaf — COMMENTED** [review]
Can you rebase and add these functions to the documentation?

File: python/src/equistore/operations/slice.py [hunk 22]
```diff
@@ -0,0 +1,395 @@
...
+    # Generate arrays of bools indicating which samples indices to keep upon slicing.
+    if samples_to_slice is None:
+        # No samples slicing wanted; keep all properties indices.
+        samples_filter = np.array([True] * len(tensorblock.samples))
+    else:
+        samples_filter = np.array(
+            [
+                sample in samples_to_slice.tolist()
+                for sample in tensorblock.samples[list(samples_to_slice.names)].tolist()
+            ]
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:310]: I'm wondering how expensive this and the other list comprehension are going to get. The main issue I have is that this is going to be O(n^2) (`sample in ...` does a full loop search), so it might be better to use specialized data structures such as a set (converting `samples_to_slice.tolist()` to a set)

Comment (ceriottm) [reply python/src/equistore/operations/slice.py:310]: Intervening just to say that indeed it'd be good to make sure that these utility functions are not going to become unexpected computational bottlenecks further down the line.

Comment (jwa7) [reply python/src/equistore/operations/slice.py:310]: Understood, have talked to Guillaume and will have a go at some line profiling and optimisation.

File: python/src/equistore/operations/slice.py (hunk 6 above)
Comment (jwa7) [reply python/src/equistore/operations/slice.py]: This, and other docstring-related changes also suggested in this comment, are fixed in ae803ab9a838accd30dc9e84c02839014996c4fc.

File: python/src/equistore/operations/slice.py (hunk 17 above)
Comment (jwa7) [reply python/src/equistore/operations/slice.py]: This and other AssertionError-related changes fixed in 993cd37389e1904cd93395d529ed452d0a3c6b4a.

File: python/src/equistore/operations/slice.py (hunk 16 above)
Comment (jwa7) [reply python/src/equistore/operations/slice.py]: Fixed in 53647819c7f11bf81ced9098208286d9ca537947.

File: python/src/equistore/operations/slice.py [hunk 23]
```diff
@@ -327,21 +327,30 @@ def _slice_tensorblock(
     # create filters for Gradient TensorBlocks as there isn't a one-to-one mapping
     # between samples of the parent TensorBlock and its Gradients. This doesn't
     # need to be done for properties as the mapping is by definition a one-to-one.
-    numeric_idxs = np.array([i for i, val in enumerate(samples_filter) if val])
+    numeric_idxs = set(np.array([i for i, val in enumerate(samples_filter) if val]))
```
Comment (Luthaf) [inline python/src/equistore/operations/slice.py:330]: I don't think you need to construct the np.array here
```suggestion
    numeric_idxs = set([i for i, val in enumerate(samples_filter) if val])
```

Comment (Luthaf) [reply python/src/equistore/operations/slice.py:330]: Even better, you can use a generator to construct the set and also bypass the list:
```suggestion
    numeric_idxs = set(i for i, val in enumerate(samples_filter) if val)
```

Comment (jwa7) [reply python/src/equistore/operations/slice.py]: Change made in 294420dcdbfab3649434b39052e91ca10aaa9c61.

**Comment by jwa7** [comment]
Recent commit 294420dcdbfab3649434b39052e91ca10aaa9c61 submitted with some more optimisations and clearing up. By my calculations using 'line_profiler' (https://github.com/pyutils/line_profiler) the code is roughly 50% faster now than before commit c22c2de.

**Comment by jwa7** [comment]
Output from `line_profiler` when running 10k calls on the `_slice_tensorblock` function:

```
Total time: 3.25904 s
File: /opt/miniforge3/envs/dev/lib/python3.10/site-packages/equistore/operations/slice.py
Function: _slice_tensorblock at line 265

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   265                                           @profile
   266                                           def _slice_tensorblock(
   267                                               tensorblock: TensorBlock, samples_to_slice=None, properties_to_slice=None
   268                                           ) -> TensorBlock:
   269                                               """
   270                                               Slices an input :py:class:`TensorBlock` along the samples and/or properties
   271                                               dimension(s). ``samples_to_slice`` and ``properties_to_slice`` are
   272                                               :py:class:`Labels` objects that specify the samples/properties
   273                                               (respectively) names and indices that should be sliced, i.e. kept in the
   274                                               output :py:class:`TensorBlock`.
   275                                           
   276                                               Note that either ``samples_to_slice`` or ``properties_to_slice``, or both,
   277                                               should be specified as input.
   278                                           
   279                                               :param tensorblock: the input :py:class:`TensorBlock` to be sliced.
   280                                               :param samples_to_slice: a :py:class:`Labels` object containing the names
   281                                                   and indices of samples to keep in the sliced output
   282                                                   :py:class:`TensorBlock`. Default value of None indicates no slicing
   283                                                   along the samples dimension should occur.
   284                                               :param properties_to_slice: a :py:class:`Labels` object containing the names
   285                                                   and indices of properties to keep in the sliced output
   286                                                   :py:class:`TensorBlock`. Default value of None indicates no slicing
   287                                                   along the properties dimension should occur.
   288                                           
   289                                               :return new_block: a :py:class:`TensorBlock` that corresponds to the sliced
   290                                                   input.
   291                                               """
   292                                           
   293     10000      38051.0      3.8      1.2      new_values = tensorblock.values
   294     10000     109715.0     11.0      3.4      new_samples = tensorblock.samples
   295     10000     129369.0     12.9      4.0      new_properties = tensorblock.properties
   296                                           
   297                                               # Generate arrays of bools indicating which samples indices to keep upon slicing.
   298     10000       1342.0      0.1      0.0      if samples_to_slice is not None:
   299     10000     124289.0     12.4      3.8          samples = tensorblock.samples[list(samples_to_slice.names)].tolist()
   300     10000      20075.0      2.0      0.6          set_samples_to_slice = set(samples_to_slice.tolist())
   301     10000      19727.0      2.0      0.6          samples_filter = np.array(
   302     10000      40610.0      4.1      1.2              [sample in set_samples_to_slice for sample in samples]
   303                                                   )
   304     10000      18770.0      1.9      0.6          new_values = new_values[samples_filter]
   305     10000      16269.0      1.6      0.5          new_samples = new_samples[samples_filter]
   306                                           
   307                                               # Generate array of bools indicating which properties indices to keep upon slicing.
   308     10000       1363.0      0.1      0.0      if properties_to_slice is not None:
   309                                                   properties = tensorblock.properties[list(properties_to_slice.names)].tolist()
   310                                                   set_properties_to_slice = set(properties_to_slice.tolist())
   311                                                   properties_filter = np.array(
   312                                                       [prop in set_properties_to_slice for prop in properties]
   313                                                   )
   314                                                   new_values = new_values[..., properties_filter]
   315                                                   new_properties = new_properties[properties_filter]
   316                                           
   317                                               # Create a new TensorBlock, sliced along the samples and properties dimension.
   318     10000     330386.0     33.0     10.1      new_block = TensorBlock(
   319     10000        802.0      0.1      0.0          values=new_values,
   320     10000        753.0      0.1      0.0          samples=new_samples,
   321     10000     138150.0     13.8      4.2          components=tensorblock.components,
   322     10000       1235.0      0.1      0.0          properties=new_properties,
   323                                               )
   324                                           
   325                                               # Create a list of numeric indices of the samples that were sliced. Needed to
   326                                               # create filters for Gradient TensorBlocks as there isn't a one-to-one mapping
   327                                               # between samples of the parent TensorBlock and its Gradients. This doesn't
   328                                               # need to be done for properties as the mapping is by definition a one-to-one.
   329     10000       1285.0      0.1      0.0      if samples_to_slice is not None:
   330     10000      75015.0      7.5      2.3          set_samples_idxs_to_slice = set(
   331     10000       2130.0      0.2      0.1              i for i, val in enumerate(samples_filter) if val
   332                                                   )
   333                                           
   334                                               # Slice each Gradient TensorBlock and add to the new_block.
   335     20000      84326.0      4.2      2.6      for parameter, gradient in tensorblock.gradients():
   336                                           
   337     20000      73968.0      3.7      2.3          new_grad_data = gradient.data
   338     20000     211577.0     10.6      6.5          new_grad_samples = gradient.samples
   339                                           
   340                                                   # Create a samples filter for the Gradient TensorBlock
   341     20000       2711.0      0.1      0.1          if samples_to_slice is not None:
   342     20000     216453.0     10.8      6.6              grad_samples = gradient.samples["sample"].tolist()
   343     20000      61552.0      3.1      1.9              grad_samples_filter = np.array(
   344     20000      98414.0      4.9      3.0                  [grad_samp in set_samples_idxs_to_slice for grad_samp in grad_samples]
   345                                                       )
   346     20000      38751.0      1.9      1.2              new_grad_samples = new_grad_samples[grad_samples_filter]
   347     20000      42666.0      2.1      1.3              new_grad_data = new_grad_data[grad_samples_filter]
   348     20000       2384.0      0.1      0.1          if properties_to_slice is not None:
   349                                                       new_grad_data = new_grad_data[..., properties_filter]
   350                                           
   351                                                   # Add sliced Gradient to the TensorBlock
   352     20000     812061.0     40.6     24.9          new_block.add_gradient(
   353     20000       1519.0      0.1      0.0              parameter=parameter,
   354     20000       1436.0      0.1      0.0              samples=new_grad_samples,
   355     20000     538332.0     26.9     16.5              components=gradient.components,
   356     20000       2178.0      0.1      0.1              data=new_grad_data,
   357                                                   )
   358                                           
   359     10000       1375.0      0.1      0.0      return new_block
```

Most of the % time is spent on assigning values tensors to variables, creating new blocks and adding gradients, which are (I think) are unavoidable parts of the method.

When no gradients are involved, the computation time for 10k function calls drops to 1.05 s. Output:

```
Total time: 1.05403 s
File: /opt/miniforge3/envs/dev/lib/python3.10/site-packages/equistore/operations/slice.py
Function: _slice_tensorblock at line 265

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   265                                           @profile
   266                                           def _slice_tensorblock(
   267                                               tensorblock: TensorBlock, samples_to_slice=None, properties_to_slice=None
   268                                           ) -> TensorBlock:
   269                                               """
   270                                               Slices an input :py:class:`TensorBlock` along the samples and/or properties
   271                                               dimension(s). ``samples_to_slice`` and ``properties_to_slice`` are
   272                                               :py:class:`Labels` objects that specify the samples/properties
   273                                               (respectively) names and indices that should be sliced, i.e. kept in the
   274                                               output :py:class:`TensorBlock`.
   275                                           
   276                                               Note that either ``samples_to_slice`` or ``properties_to_slice``, or both,
   277                                               should be specified as input.
   278                                           
   279                                               :param tensorblock: the input :py:class:`TensorBlock` to be sliced.
   280                                               :param samples_to_slice: a :py:class:`Labels` object containing the names
   281                                                   and indices of samples to keep in the sliced output
   282                                                   :py:class:`TensorBlock`. Default value of None indicates no slicing
   283                                                   along the samples dimension should occur.
   284                                               :param properties_to_slice: a :py:class:`Labels` object containing the names
   285                                                   and indices of properties to keep in the sliced output
   286                                                   :py:class:`TensorBlock`. Default value of None indicates no slicing
   287                                                   along the properties dimension should occur.
   288                                           
   289                                               :return new_block: a :py:class:`TensorBlock` that corresponds to the sliced
   290                                                   input.
   291                                               """
   292                                           
   293     10000      36321.0      3.6      3.4      new_values = tensorblock.values
   294     10000     105538.0     10.6     10.0      new_samples = tensorblock.samples
   295     10000     125617.0     12.6     11.9      new_properties = tensorblock.properties
   296                                           
   297                                               # Generate arrays of bools indicating which samples indices to keep upon slicing.
   298     10000       1312.0      0.1      0.1      if samples_to_slice is not None:
   299     10000     121562.0     12.2     11.5          samples = tensorblock.samples[list(samples_to_slice.names)].tolist()
   300     10000      18851.0      1.9      1.8          set_samples_to_slice = set(samples_to_slice.tolist())
   301     10000      19196.0      1.9      1.8          samples_filter = np.array(
   302     10000      35912.0      3.6      3.4              [sample in set_samples_to_slice for sample in samples]
   303                                                   )
   304     10000      17520.0      1.8      1.7          new_values = new_values[samples_filter]
   305     10000      14978.0      1.5      1.4          new_samples = new_samples[samples_filter]
   306                                           
   307                                               # Generate array of bools indicating which properties indices to keep upon slicing.
   308     10000       1358.0      0.1      0.1      if properties_to_slice is not None:
   309                                                   properties = tensorblock.properties[list(properties_to_slice.names)].tolist()
   310                                                   set_properties_to_slice = set(properties_to_slice.tolist())
   311                                                   properties_filter = np.array(
   312                                                       [prop in set_properties_to_slice for prop in properties]
   313                                                   )
   314                                                   new_values = new_values[..., properties_filter]
   315                                                   new_properties = new_properties[properties_filter]
   316                                           
   317                                               # Create a new TensorBlock, sliced along the samples and properties dimension.
   318     10000     329152.0     32.9     31.2      new_block = TensorBlock(
   319     10000        811.0      0.1      0.1          values=new_values,
   320     10000        745.0      0.1      0.1          samples=new_samples,
   321     10000     134810.0     13.5     12.8          components=tensorblock.components,
   322     10000       1124.0      0.1      0.1          properties=new_properties,
   323                                               )
   324                                           
   325                                               # Create a list of numeric indices of the samples that were sliced. Needed to
   326                                               # create filters for Gradient TensorBlocks as there isn't a one-to-one mapping
   327                                               # between samples of the parent TensorBlock and its Gradients. This doesn't
   328                                               # need to be done for properties as the mapping is by definition a one-to-one.
   329     10000       1268.0      0.1      0.1      if samples_to_slice is not None:
   330     10000      67687.0      6.8      6.4          set_samples_idxs_to_slice = set(
   331     10000       1925.0      0.2      0.2              i for i, val in enumerate(samples_filter) if val
   332                                                   )
   333                                           
   334                                               # Slice each Gradient TensorBlock and add to the new_block.
   335     10000      17038.0      1.7      1.6      for parameter, gradient in tensorblock.gradients():
   336                                           
   337                                                   new_grad_data = gradient.data
   338                                                   new_grad_samples = gradient.samples
   339                                           
   340                                                   # Create a samples filter for the Gradient TensorBlock
   341                                                   if samples_to_slice is not None:
   342                                                       grad_samples = gradient.samples["sample"].tolist()
   343                                                       grad_samples_filter = np.array(
   344                                                           [grad_samp in set_samples_idxs_to_slice for grad_samp in grad_samples]
   345                                                       )
   346                                                       new_grad_samples = new_grad_samples[grad_samples_filter]
   347                                                       new_grad_data = new_grad_data[grad_samples_filter]
   348                                                   if properties_to_slice is not None:
   349                                                       new_grad_data = new_grad_data[..., properties_filter]
   350                                           
   351                                                   # Add sliced Gradient to the TensorBlock
   352                                                   new_block.add_gradient(
   353                                                       parameter=parameter,
   354                                                       samples=new_grad_samples,
   355                                                       components=gradient.components,
   356                                                       data=new_grad_data,
   357                                                   )
   358                                           
   359     10000       1302.0      0.1      0.1      return new_block
```

**Review by Luthaf — APPROVED** [review]
Looks good! Can you rebase & squash your commits;  and add the function to the documentation (it is a manual process since #78)?

**Comment by jwa7** [comment]
Changes made to unit test imports, functions added to docs, commits squashed and changes force pushed

File: docs/src/reference/python/operations/slice.rst [hunk 24]
```diff
@@ -0,0 +1,6 @@
+slice
```
Comment (PicoCentauri) [inline docs/src/reference/python/operations/slice.rst:1]: Could you just capitalize this here. Just to be consistent,

```suggestion
Slice
```

File: python/src/equistore/operations/slice.py [hunk 25]
```diff
@@ -0,0 +1,399 @@
...
+    Example: take an input :py:class:`TensorBlock` of shape (100, 1, 6), where
+    there are 100 'samples', 1 'components', and 6 'properties'. Say we want to
+    slice this tensor along the samples and properties dimensions. As in the
+    code-block below, we can specify, for example, 4 samples and 3 properties
+    indices to keep. The returned :py:class:`TensorBlock` will have shape (4, 1,
+    3).
+
+    .. code-block:: python
+    
+        sliced_tensorblock = _slice_tensorblock(
```
Comment (PicoCentauri) [inline python/src/equistore/operations/slice.py:171]: Why do refer to `_slice_tensorblock ` here instead of giving an example to the function you are documenting?

Comment (jwa7) [reply python/src/equistore/operations/slice.py:171]: Good point thank you!

File: docs/src/reference/python/operations/slice.rst (hunk 24 above)
Comment (jwa7) [reply docs/src/reference/python/operations/slice.rst:1]: Consistent with what? Granted there is capitalisation in index.rst and samples_reduction.rst, but not in allclose.rst, dot.rst, lstsq.rst, remove_gradients.rst, or solve.rst

File: python/src/equistore/operations/slice.py (hunk 25 above)
Comment (jwa7) [reply python/src/equistore/operations/slice.py]: Change made in 1f8f8b8c81e91af663db59a154967c9c9707e186.

File: docs/src/reference/python/operations/slice.rst (hunk 24 above)
Comment (PicoCentauri) [reply docs/src/reference/python/operations/slice.rst:1]: Ahh okay we can also leave it. I just saw the header of `Operations`
